File: identify_chunks/fastcdc.c
//
// Created by borelset on 2019/3/20.
//

#include <stdint.h>
#include <memory.h>
#include <assert.h>
#include <stdio.h>
#include "md5.h"

#define SymbolCount 256
#define DigistLength 16
#define SeedLength 64
#define MaxChunkSizeOffset 3
#define MinChunkSizeOffset 2

uint64_t g_gear_matrix[SymbolCount];
uint32_t g_min_fastcdc_chunk_size;
uint32_t g_max_fastcdc_chunk_size;
uint32_t g_expect_fastcdc_chunk_size;

enum{
    Mask_64B,
    Mask_128B,
    Mask_256B,
    Mask_512B,
    Mask_1KB,
    Mask_2KB,
    Mask_4KB,
    Mask_8KB,
    Mask_16KB,
    Mask_32KB,
    Mask_64KB,
    Mask_128KB
};

uint64_t g_condition_mask[] = {
        0x00001803110,// 64B
        0x000018035100,// 128B
        0x00001800035300,// 256B
        0x000019000353000,// 512B
        0x0000590003530000,// 1KB
        0x0000d90003530000,// 2KB
        0x0000d90103530000,// 4KB
        0x0000d90303530000,// 8KB
        0x0000d90313530000,// 16KB
        0x0000d90f03530000,// 32KB
        0x0000d90303537000,// 64KB
        0x0000d90703537000// 128KB
};

void fastcdc_init(uint32_t expectCS){
    char seed[SeedLength];
    for(int i=0; i<SymbolCount; i++){
        for(int j=0; j<SeedLength; j++){
            seed[j] = i;
        }

        g_gear_matrix[i] = 0;
        char md5_result[DigistLength];
        md5_state_t md5_state;
        md5_init(&md5_state);
        md5_append(&md5_state, seed, SeedLength);
        md5_finish(&md5_state, md5_result);

        memcpy(&g_gear_matrix[i], md5_result, sizeof(uint64_t));
    }

    g_min_fastcdc_chunk_size = 6144;
    g_max_fastcdc_chunk_size = 65536;
    g_expect_fastcdc_chunk_size = expectCS;
}


int fastcdc_chunk_data(unsigned char *p, int n){

    uint64_t fingerprint=0,digest;
    int i=g_min_fastcdc_chunk_size, Mid=g_min_fastcdc_chunk_size + 4*1024;
    //return n;

    if(n<=g_min_fastcdc_chunk_size) //the minimal  subChunk Size.
        return n;
    //windows_reset();
    if(n > g_max_fastcdc_chunk_size)
        n = g_max_fastcdc_chunk_size;
    else if(n<Mid)
        Mid = n;
    while(i<Mid)
    {
        fingerprint = (fingerprint<<1) + (g_gear_matrix[p[i]]);
        if ((!(fingerprint & 0x0000d90f03530000))) { //AVERAGE*2, *4, *8
            return i;
        }
        i++;
    }
    while(i<n   )
    {
        fingerprint = (fingerprint<<1) + (g_gear_matrix[p[i]]);
        if ((!(fingerprint & 0x0000d90003530000))) { //Average/2, /4, /8
            return i;
        }
        i++;
    }
    //printf("\r\n==chunking FINISH!\r\n");
    return i;

    /*
    assert(p != NULL);
    uint64_t fp = 0;
    if(n < g_max_fastcdc_chunk_size){
        if(n <= g_min_fastcdc_chunk_size){
            return n;
        }else if (n < g_expect_fastcdc_chunk_size){
            for(int i=g_min_fastcdc_chunk_size; i<n; i++){
                fp = ( fp <<1 ) + g_gear_matrix[ p[i] ];
                if(!(fp & g_condition_mask[Mask_64KB])){
                    return i;
                }
            }
            return n;
        }else {
            for(int i=g_min_fastcdc_chunk_size; i<g_expect_fastcdc_chunk_size; i++){
                fp = ( fp <<1 ) + g_gear_matrix[ p[i] ];
                if(!(fp & g_condition_mask[Mask_64KB])){
                    return i;
                }
            }
            for(int i=g_expect_fastcdc_chunk_size; i<n; i++){
                fp = ( fp <<1 ) + g_gear_matrix[ p[i] ];
                if(!(fp & g_condition_mask[Mask_1KB])){
                    return i;
                }
            }
            return n;
        }
    }
    for(int i=g_min_fastcdc_chunk_size; i<g_expect_fastcdc_chunk_size; i++){
        fp = ( fp <<1 ) + g_gear_matrix[ p[i] ];
        if(!(fp & g_condition_mask[Mask_64KB])){
            return i;
        }
    }
    for(int i=g_expect_fastcdc_chunk_size; i<g_max_fastcdc_chunk_size; i++){
        fp = ( fp <<1 ) + g_gear_matrix[ p[i] ];
        if(!(fp & g_condition_mask[Mask_1KB])){
            return i;
        }
    }
    return g_max_fastcdc_chunk_size;
     */
}

File: identify_chunks/fastcdc.js
var hash123;
if(!hash123)
{
    hash123 = require('./hash123');
}

var fastcdc = new function()
{
    var SymbolCount = 256;
    var DigistLength = 16;
    var SeedLength = 64;
    var MaxChunkSizeOffset = 1;
    var MinChunkSizeOffset = 2;

    var g_gear_matrix = new ArrayBuffer(SymbolCount*32);
    //g_gear_matrix_view64 = new BigUint64Array(g_gear_matrix);
    var g_gear_matrix_view32 = new Uint32Array(g_gear_matrix);
    var g_min_chunk_size = 0;
    var g_max_chunk_size = 0;
    var minMask;
    var maxMask;
    var EBlockLength;

    var mask = {
        'Mask_16B': 0x000080010300,
        'Mask_32B': 0x000080050300,
        'Mask_64B': 0x000080070300,
        'Mask_128B': 0x000090100053,
        'Mask_256B': 0x000091300301,
        'Mask_512B': 0x000090055013,
        'Mask_1KB': 0x000083030531,
        'Mask_2KB': 0x0000d9000353,
        'Mask_4KB': 0x0000d9010353,
        'Mask_8KB': 0x0000d9030353,
        'Mask_16KB': 0x0000d9070353,
        'Mask_32KB': 0x0000f9070353,
        'Mask_64KB': 0x0000f90f0353,
        'Mask_128KB': 0x0000f90f0373
    };
    var BlockLen = {
        'Mask_16B': 16,
        'Mask_32B': 32,
        'Mask_64B': 64,
        'Mask_128B': 128,
        'Mask_256B': 256,
        'Mask_512B': 512,
        'Mask_1KB': 1024,
        'Mask_2KB': 2048,
        'Mask_4KB': 4096,
        'Mask_8KB': 8192,
        'Mask_16KB': 16384,
        'Mask_32KB': 32768,
        'Mask_64KB': 65536,
        'Mask_128KB': 131072
    };


    function Init(expectCS)
    {
        seed = new ArrayBuffer(SeedLength) ;
        seedView = new Uint8Array(seed);

        for(i=0; i<SymbolCount; i++){
            for(j=0; j<SeedLength; j++){
                seedView[j] = i;
            }
            g_gear_matrix[i] = 0;

            md5_result = hash123.md5(seedView,0,0,SeedLength);
            g_gear_matrix_view32[i] = md5_result[0];
        }
        if (expectCS == 'Mask_32B')
        {
            minMask = mask['Mask_16B'];
            maxMask = mask['Mask_64B'];
        }
        else if (expectCS == 'Mask_64B')
        {
            minMask = mask['Mask_32B'];
            maxMask = mask['Mask_128B'];
        }
        else if (expectCS == 'Mask_128B')
        {
            minMask = mask['Mask_64B'];
            maxMask = mask['Mask_256B'];
        }
        else if (expectCS == 'Mask_256B')
        {
            minMask = mask['Mask_128B'];
            maxMask = mask['Mask_512B'];
        }
        else if (expectCS == 'Mask_512B')
        {
            minMask = mask['Mask_256B'];
            maxMask = mask['Mask_1KB'];
        }
        else if (expectCS == 'Mask_1KB')
        {
            minMask = mask['Mask_512B'];
            maxMask = mask['Mask_2KB'];
        }
        else if (expectCS == 'Mask_2KB')
        {
            minMask = mask['Mask_1KB'];
            maxMask = mask['Mask_4KB'];
        }
        else if (expectCS == 'Mask_4KB')
        {
            minMask = mask['Mask_2KB'];
            maxMask = mask['Mask_8KB'];
        }
        else if (expectCS == 'Mask_8KB')
        {
            minMask = mask['Mask_4KB'];
            maxMask = mask['Mask_16KB'];
        }
        else if (expectCS == 'Mask_16KB')
        {
            minMask = mask['Mask_8KB'];
            maxMask = mask['Mask_32KB'];
        }
        else if (expectCS == 'Mask_32KB')
        {
            minMask = mask['Mask_16KB'];
            maxMask = mask['Mask_64KB'];
        }

        EBlockLength = BlockLen[expectCS];

    }


    // function ChunkData(dataView, start, length){
    //     var fp = 0;
    //     for(var i=start; i<length; i++) {
    //         fp = (fp << 1) + g_gear_matrix_view32[dataView[i]];
    //         if (!(fp & Mask)) {
    //             return [i + 1, (fp & 0xffffffff)];
    //         }
    //     }
    //     return [i, (fp&0xffffffff)];
    // }

    function ChunkData(dataView, start, length){
        var fp = 0;
        var gearHash = 0;
        var Mask = maxMask;

        for(var i=start; i<length; i++) {
            if((i - start) == EBlockLength);
                Mask = minMask;

            if(((i-start) & 0x0000000f) == 0 )
            {
                gearHash += fp;
            }
            fp = (fp << 1) ^ g_gear_matrix_view32[dataView[i]];

            if (!(fp & Mask)) {
                var fastFp =  gearHash + fp ;
                return [i + 1, fastFp];
            }
        }

        fastFp =  gearHash + fp ;
        return [i, fastFp];
    }

    this.Init = Init;
    this.ChunkData = ChunkData;
}

if(((typeof require) != "undefined") &&
    ((typeof module) != "undefined") &&
    ((typeof module.exports) != "undefined"))
    module.exports = fastcdc;